system/jlib/jqueue.hpp

```diff
@@ -0,0 +1,361 @@
+                    unsigned curDequeueSeq = (curState >> dequeueShift); // No need to mask since the top field
+                    unsigned slotMask = (fixedSlotBits ? fixedSlotMask : dynamicSlotMask);
+                    unsigned expectedSeq = (curDequeueSeq + 1) & sequenceMask;
+                    unsigned curDequeueSlot = (curDequeueSeq & slotMask);
+                    BufferElement & cur = values[curDequeueSlot];
+                    loop
+                    {
+                        unsigned sequence = cur.sequence.load(std::memory_order_acquire);
+                        if (sequence == expectedSeq)
+                            break;
+                        //possibly yield every n iterations?
+                        spinPause();
+                    }
+
+                    result = std::move(cur.value);
+                    const unsigned numSlots = slotMask + 1;
+                    unsigned nextSeq = (curDequeueSeq + numSlots) & sequenceMask;
+                    cur.sequence.store(nextSeq, std::memory_order_release);
+                    if ((curState & writerMask) != 0)
+                        writers.signal();
+                    return true;
+                }
+            }
+        }
+    }
+
+    virtual void reset()
+    {
+        activeWriters.store(maxWriters, std::memory_order_relaxed);
+        aborted.store(false, std::memory_order_relaxed);
+        readers.reinit(0);
+        writers.reinit(0);
+    }
+    virtual void noteWriterStopped()
+    {
+        //MORE: If this reduces activeProducers to 0 then it may need to wake up any waiting threads.
+        if (--activeWriters <= 0)
+        {
+            state_t curState = state.load(std::memory_order_acquire);
+            unsigned readersWaiting = (curState & readerMask) >> readerShift;
+            readers.signal(readersWaiting);
+        }
+    }
+    virtual void abort()
+    {
+        //readers and writers may enqueue/dequeue another row before this takes effect
+        aborted.store(true, std::memory_order_release);
+        state_t curState = state.load(std::memory_order_acquire);
+        unsigned readersWaiting = (curState & readerMask) >> readerShift;
+        unsigned writersWaiting = (curState & writerMask) >> writerShift;
+        readers.signal(readersWaiting);
+        writers.signal(writersWaiting);
+    }
+    inline bool allWritersStopped() const { return activeWriters.load(std::memory_order_acquire) <= 0; }
+
+protected:
+    BufferElement * values;
+    unsigned dynamicSlotMask;
+    unsigned maxItems;
+    unsigned maxWriters;
+    std::atomic<int> activeWriters;
+    std::atomic<bool> aborted;
+    char pad1[64];
+    Semaphore readers;
+    char pad2[64-sizeof(Semaphore)];
+    Semaphore writers;
+    char pad3[64-sizeof(Semaphore)];
+    //Ensure the state is not on the same cache line as anything else, especially anything that is modified.
+    std::atomic<state_t> state;
+};
+
+typedef IQueueOf<const void *> IRowQueue;
```

Comment:
Are these suitable for general use or are they pretty specific / user needs to know what they are doing?

If the latter may want a different name

Either way some comments in here on how this interface/object is designed to be used might be handy?